### PR TITLE
Get remote objects by commit hash

### PR DESCRIPTION
**Pull request by mariusraht2**
Related to #3040 Revert to any single commit

Get remote objects by commit hash not only by branch name and updated method calls accordingly if neccessary,

### CODE REVIEW

File: src/git/zcl_abapgit_git_transport.clas.abap [hunk 1]
```diff
@@ -7,17 +7,16 @@ CLASS zcl_abapgit_git_transport DEFINITION
...
+      IMPORTING iv_url          TYPE string
+                iv_branch_name  TYPE string
+                iv_commit_hash  TYPE zif_abapgit_definitions=>ty_sha1 OPTIONAL
+                iv_deepen_level TYPE numc2 DEFAULT '0'
+                it_branches     TYPE zif_abapgit_definitions=>ty_git_branch_list_tt OPTIONAL
+      EXPORTING et_objects      TYPE zif_abapgit_definitions=>ty_objects_tt
+                ev_branch       TYPE zif_abapgit_definitions=>ty_sha1
+                eo_branch_list  TYPE REF TO zcl_abapgit_git_branch_list
+                ev_commit       TYPE zif_abapgit_definitions=>ty_sha1
+      RAISING   zcx_abapgit_exception.
```
Comment (larshp) [inline src/git/zcl_abapgit_git_transport.clas.abap:19]: please reformat so it follows the old code

File: src/git/zcl_abapgit_git_transport.clas.abap [hunk 2]
```diff
@@ -7,17 +7,16 @@ CLASS zcl_abapgit_git_transport DEFINITION
...
-      EXPORTING
-        !et_objects     TYPE zif_abapgit_definitions=>ty_objects_tt
-        !ev_branch      TYPE zif_abapgit_definitions=>ty_sha1
-        !eo_branch_list TYPE REF TO zcl_abapgit_git_branch_list
-      RAISING
-        zcx_abapgit_exception .
+      IMPORTING iv_url          TYPE string
+                iv_branch_name  TYPE string
+                iv_commit_hash  TYPE zif_abapgit_definitions=>ty_sha1 OPTIONAL
+                iv_deepen_level TYPE numc2 DEFAULT '0'
```
Comment (larshp) [inline src/git/zcl_abapgit_git_transport.clas.abap:13]: what is the background for changing the logic regarding deepen? This PR concerns getting remote objects by commit hash, so I do not expect the deepen logic to change

Comment (mariusraht2) [reply src/git/zcl_abapgit_git_transport.clas.abap:13]: I call the git api with deepen_level = 99 to select as much commit hashes as possible and after that load them into the popup to be displayed to be able to select one one want to revert to. In my first concept I tried to get them by its hierarchie but unfortunately sometimes commits seem to have no parent I would know at one step in the hierarchie so I would have to select "blindly" to step down the history.

**Comment by larshp** [comment]
input parameter `iv_commit_hash` has been added to the method, but it already has `branch_name` which as far as I remember refers to the same thing?

**Comment by mariusraht2** [comment]
> input parameter `iv_commit_hash` has been added to the method, but it already has `branch_name` which as far as I remember refers to the same thing?
Kind of: In the first option we use branch_name to determine the branch hash and in the second option we use the imported commit hash to get what we want. I suggest we should split the functions into separate methods - it's sort of confusing what to supply and the method already does too much in my optinion. I think in the long run we should change the methods which are using just branch hashes to get development objects (like upload_pack) to import hashes directly (instead of additionally calling find_branch or any other determination logic).

**Comment by larshp** [comment]
sounds good, please create a new method for the new functionality, new method name `upload_pack_by_hash`

we can rename the old method to `upload_pack_by_branch` in a follow up PR later

**Comment by mariusraht2** [comment]
Don't try it out yet - I need to fix some errors related to the changes I made.

**Comment by mariusraht2** [comment]
@larshp Ok, now it's ready - please see the commit description. I hope the changes are ok - don't like redundant source code so I reused the method upload_pack (see description of the latest commit).

File: src/git/zcl_abapgit_git_transport.clas.abap [hunk 3]
```diff
@@ -6,16 +6,27 @@ CLASS zcl_abapgit_git_transport DEFINITION
...
+        !et_objects      TYPE zif_abapgit_definitions=>ty_objects_tt
+        !ev_branch       TYPE zif_abapgit_definitions=>ty_sha1
+        !eo_branch_list  TYPE REF TO zcl_abapgit_git_branch_list
+      RAISING
+        zcx_abapgit_exception .
+    CLASS-METHODS upload_pack_by_hash
+      IMPORTING
+        !iv_url          TYPE string
+        !iv_branch_name  TYPE string
+        !iv_hash         TYPE zif_abapgit_definitions=>ty_sha1 OPTIONAL
```
Comment (larshp) [inline src/git/zcl_abapgit_git_transport.clas.abap:25]: what is the actual hash to be supplied here? is it the tree hash, the commit hash, the blob hash? in any case I don't think the branch name is required along with the hash?

File: src/git/zcl_abapgit_git_transport.clas.abap [hunk 4]
```diff
@@ -6,16 +6,27 @@ CLASS zcl_abapgit_git_transport DEFINITION
...
-    CLASS-METHODS upload_pack
+    CLASS-METHODS upload_pack_by_branch
       IMPORTING
-        !iv_url         TYPE string
-        !iv_branch_name TYPE string
-        !iv_deepen      TYPE abap_bool DEFAULT abap_true
-        !it_branches    TYPE zif_abapgit_definitions=>ty_git_branch_list_tt OPTIONAL
+        !iv_url          TYPE string
+        !iv_branch_name  TYPE string
+        !iv_deepen_level TYPE numc2 DEFAULT 0
```
Comment (larshp) [inline src/git/zcl_abapgit_git_transport.clas.abap:13]: earlier the default behavior was
if iv_deepen was true then the depth was 1, else everything
now the default behavior is get everything?

please roll this back to the old behavior, this change can be introduced in a follow up PR

**Comment by larshp** [comment]
> we can rename the old method to `upload_pack_by_branch` in a follow up PR later

and ^

File: src/git/zcl_abapgit_git_transport.clas.abap (hunk 3 above)
Comment (mariusraht2) [reply src/git/zcl_abapgit_git_transport.clas.abap:25]: The method is able to handle every hash type. Unfortunately we need the branch name to get the client object to call set_headers for the git function :(

**Comment by mariusraht2** [comment]
@larshp Changed related to your review.

**Comment by larshp** [comment]
A: Can we please keep the deepen logic to the old behavior for now? It can be changed later after the new method is added, this makes it a lot easier to review this PR
B: Please try to leave out whitespace changes, it makes it a lot easier to review. Right now file zcl_abapgit_git_porcelain is changed only with whitespace changes
C: Getting the client object, that is internal logic to the class. No need to require fields externally that are not really needed. First step is to find the right signature for the method.
D: right now METHOD upload_pack_by_hash is largely a copy of the old method, this needs refactoring to avoid duplication

**Comment by mariusraht2** [comment]
A) Ok.
B) I'll try.
C) I'll take a look into it - maybe I can avoid the call or find a better object.. don't know yet.
D) Yes, that's why I renamed update_pack to update_pack_by_branch and then reused upload_pack in the previous review. But you wrote that we should rename upload_pack to upload_pack_by_branch in a later PR so I just didn't try to move duplicate source code to an extra method (maybe also in a later PR?).

**Comment by mariusraht2** [comment]
I suggest the following:
1) I'll create a new PR just for renaming upload_pack to upload_pack_by_branch and updating the calls respectivly.
2) After that I'll create a PR for adding a new method upload_pack_by_commit and readding the upload_pack method with the duplicate source code upload:_pack_by_branch and upload_pack_by_commit share.
3) Then I'll create a PR for changing the deepen_level.
4) The 4th PR would be for renaming pull to pull_by_branch
5) The 5th PR would add a new method pull_by_commit

What do you think?

**Comment by mariusraht2** [comment]
I did some additional research related to the method upload_pack: Unfortunately I have to call find_branch also in the upload_pack_by_commit because I need the branches hash also for the following logic. So in the end the only differences between upload_pack_by_branch and upload_pack_by_commit are the following:
- upload_pack_by_commit requires only an additional exporting parameter ev_commit which holds the current parent hash the git action was taken for. We still need ev_branch and thus have to determine it with the method find_branch (related to question/comment C)
- upload_pack logic is being called either for branch_hash or commit_hash and related to that lt_wanteds is being filled

It's really difficult to separate the commit and branch related methods because the methods were created with the intention to work primarily with the branch. It would have been better if they would have been designed to work with hashes.

**Comment by larshp** [comment]
first step is designing the right parameters against the callers, how the class works internally is next step

**Comment by mariusraht2** [comment]
Yes, ok, but unfortunately I don't understand yet what you would do to solve this issue. Would you create 3 methods (upload_pack_by_branch, upload_pack_by_commit + 1 for shared source code) or just 1 with an additional import like the first solution or do you favor another solution?

**Comment by larshp** [comment]
two public methods upload_pack_by_branch + upload_pack_by_commit, with inputs and outputs corresponding to the use case

the implementation is inside the class, no internal implementation restrictions or current design should make it into the 2 public method parameters

**Comment by mariusraht2** [comment]
And how do we avoid the redundant source code? Only the filling of lt_hashes is different in each of those methods:
```
IF it_branches IS INITIAL.
    APPEND ev_branch TO lt_hashes.
ELSE.
    LOOP AT it_branches ASSIGNING <ls_branch>.
      APPEND <ls_branch>-sha1 TO lt_hashes.
    ENDLOOP.
ENDIF.
```
vs.
```
APPEND iv_hash TO lt_hashes.
```

**Comment by larshp** [comment]
thats internal functinoality, lets first agree on the methods, can you help listing the code for the method importing and exporting parameters, so we got that closed?

**Comment by mariusraht2** [comment]
Ok, thanks for the patience :)

```
    CLASS-METHODS upload_pack_by_commit
      IMPORTING
        !iv_url          TYPE string
        !iv_commit_hash  TYPE zif_abapgit_definitions=>ty_sha1
        !iv_branch_name  TYPE string
        !iv_deepen_level TYPE numc2 DEFAULT 0
      EXPORTING
        !et_objects      TYPE zif_abapgit_definitions=>ty_objects_tt
        !ev_branch       TYPE zif_abapgit_definitions=>ty_sha1
      RAISING
        zcx_abapgit_exception .
```
```
    CLASS-METHODS upload_pack_by_branch
      IMPORTING
        !iv_url          TYPE string
        !iv_branch_name  TYPE string
        !iv_deepen_level TYPE numc2 DEFAULT 0
        !it_branches     TYPE zif_abapgit_definitions=>ty_git_branch_list_tt OPTIONAL
      EXPORTING
        !et_objects      TYPE zif_abapgit_definitions=>ty_objects_tt
        !ev_branch       TYPE zif_abapgit_definitions=>ty_sha1
        !eo_branch_list  TYPE REF TO zcl_abapgit_git_branch_list
      RAISING
        zcx_abapgit_exception .
```

**Comment by g-back** [comment]
Pedantic suggestion, change order of import parameters in `by_commit` so they're similar to `by_branch`
```abap
 CLASS-METHODS upload_pack_by_commit
      IMPORTING
        !iv_url          TYPE string
        !iv_branch_name  TYPE string
        !iv_deepen_level TYPE numc2 DEFAULT 0
        !iv_commit_hash  TYPE zif_abapgit_definitions=>ty_sha1 "<--
       ...
```

**Comment by larshp** [comment]
branch names points to commits, so no need for branch name as input in upload_pack_by_commit

**Comment by mariusraht2** [comment]
IMHO I would say the branch name states to which branch the commit belongs to. If we would omit the branch name we wouldn't be able to get the objects (commits are only objects, too) by the branch, because we don't know it anymore. Also I think this is a method that belongs to a branch object, but we don't treat is as this kind of method which requires this method to have an import parameter which points to the branch.

**Comment by mariusraht2** [comment]
@larshp Honestyl I don't want to improve this method in this PR - I just want to use it with one different parameter (iv_commit_hash). We should come to a solution - any other PR I can image for this feature depends on this as far as I can see.

**Comment by larshp** [comment]
closing, no progress

suggest opening a PR for renaming
zcl_abapgit_git_transport=>upload_pack(
to
zcl_abapgit_git_transport=>upload_pack_by_branch(

**Pull request closed without merging**